src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const newState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          newState[key] = action.extraData[key];
+        }
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      case 'clear':
+        for (const key in newState) {
+          delete newState[key];
+        }
+        break;
+      default:
+        break;
+    }
+    stateHistory.push({ ...newState });
```

Comment:
The current implementation is mutating the 'newState' object within the 'forEach' loop, which means that all the states pushed into 'stateHistory' will reference the same object. To prevent this, you should create a new copy of the state within each iteration of the loop right before applying any changes.